pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/groupby/DefaultGroupKeyGenerator.java

```diff
@@ -298,7 +310,87 @@ public int getCurrentGroupKeyUpperBound() {
       case ARRAY_MAP_BASED:
         return new ArrayMapBasedGroupKeyIterator();
       default:
-        throw new RuntimeException("Unsupported storage type.");
+        throw new RuntimeException("Unsupported storage type for key generator " + _storageType);
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @param keysToPurge Group keys to purge
+   */
+  @Override
+  public void purgeKeys(int[] keysToPurge) {
+    switch (_storageType) {
+      case ARRAY_BASED:
+        if (keysToPurge != null && keysToPurge.length != 0) {
+          throw new RuntimeException("Illegal operation: Purging group keys not allowed for array-based key generator");
+        } // else nothing to be done.
+        break;
+
+      case ARRAY_MAP_BASED:
+        purgeArrayBasedKeys(keysToPurge);
+        break;
+
+      case LONG_MAP_BASED:
+        purgeMapBasedKeys(keysToPurge);
+        break;
+
+        default:
+          throw new RuntimeException("Unsupported storage type for key generator " + _storageType);
+    }
+  }
+
+  /**
+   * Helper method to purge group keys that got trimmed from the group by result.
+   * Lazily builds a reverse map from id to array group key on the first call.
+   *
+   * @param groupKeys Group keys to purge
+   */
+  private void purgeArrayBasedKeys(int[] groupKeys) {
+    if (groupKeys == null || groupKeys.length == 0) {
+      return; // Nothing to purge
+    }
+
+    // Lazily build the reverse map, if/when required.
+    if (_idToGroupKey == null) {
```

Comment:
No. This is null until we decide to switch the mode. We build this map only once lazily, and after that the map is non-null (ie not built per block).